File: src/modules/audio/streamExtractor.ts
/**
 * M2 — AudioStreamExtractor
 *
 * 职责边界：
 * - ✅ 直接拷贝音频流（-c:a copy），不重编码
 * - ✅ 支持任意输入容器，输出目标容器
 * - ✅ 推送进度事件
 * - ❌ 不做格式转换（不调 -c:a libmp3lame 等）
 * - ❌ 不变更码率
 *
 * 单一职责：从视频容器中无损提取音频流。
 */

import type { AudioResult } from './formatConverter';

export interface DirectExtractParams {
  inputPath: string;
  outputPath: string;
  onProgress?: (percent: number) => void;
}

export interface DirectExtractContext {
  inputPath: string;
  outputPath: string;
  app?: unknown;
  totalSecs?: number;
}

/** 纯函数：组装 -c:a copy 的 ffmpeg 参数（不实际执行） */
export const buildDirectExtractArgs = (inputPath: string, outputPath: string): string[] => {
  return ['-y', '-i', inputPath, '-vn', '-c:a', 'copy', outputPath];
};

/** Tauri 桥接：调用后端 extract_audio（Tauri command），direct 模式 */
export const invokeDirectExtract = async (params: DirectExtractParams): Promise<AudioResult> => {
  const { invoke } = await import('@tauri-apps/api/core');
  const ext = params.outputPath.split('.').pop() || 'm4a';
  const raw = await invoke<{
    output_path: string;
    original_size: number;
    output_size: number;
    saved_bytes: number;
    processing_time_secs: number;
  }>('extract_audio', {
    inputPath: params.inputPath,
    outputPath: params.outputPath,
    format: ext,
    bitrateKbps: undefined,
    mode: 'direct',
  });
  return {
    outputPath: raw.output_path,
    originalSize: raw.original_size,
    outputSize: raw.output_size,
    savedBytes: raw.saved_bytes,
    processingTimeSecs: raw.processing_time_secs,
  };
};

/** 抽象接口：可注入 mock 实现做单测 */
export interface IAudioStreamExtractor {
  extract(ctx: DirectExtractContext): Promise<AudioResult>;
}

File: src-tauri/src/domain/audio/stream_extractor.rs
use std::process::{Command, Stdio};

use tauri::Emitter;

use crate::domain::audio::AudioOperationResult;
use crate::infrastructure::ffmpeg_manager::{get_ffmpeg_path, parse_progress, FFmpegError};

/// 构造音频流复制（无损提取）的 FFmpeg 参数列表（纯函数）
pub fn build_extract_copy_args(input_path: &str, output_path: &str) -> Vec<String> {
    vec![
        String::from("-y"),
        String::from("-i"),
        input_path.to_string(),
        String::from("-vn"),
        String::from("-c:a"),
        String::from("copy"),
        output_path.to_string(),
    ]
}

pub struct ExtractContext<'a> {
    pub input_path: &'a str,
    pub output_path: &'a str,
    pub total_secs: f64,
    pub app: &'a tauri::AppHandle,
}

pub trait AudioStreamExtractor: Send + Sync {
    fn extract(&self, ctx: ExtractContext) -> Result<AudioOperationResult, String>;
}

/// FfmpegStreamExtractor — -c:a copy 模式，无损提取
pub struct FfmpegStreamExtractor;

impl AudioStreamExtractor for FfmpegStreamExtractor {
    fn extract(&self, ctx: ExtractContext) -> Result<AudioOperationResult, String> {
        let ffmpeg = get_ffmpeg_path().ok_or_else(|| {
            FFmpegError::NotFound("ffmpeg 未找到".to_string()).to_string()
        })?;

        let start = std::time::Instant::now();
        let original_size = std::fs::metadata(ctx.input_path)
            .map_err(|e| format!("读取原文件失败: {e}"))?
            .len();

        let args = build_extract_copy_args(ctx.input_path, ctx.output_path);

        let mut child = Command::new(&ffmpeg)
            .args(&args)
            .stderr(Stdio::piped())
            .stdout(Stdio::null())
            .spawn()
            .map_err(|e| format!("启动 ffmpeg 失败: {e}"))?;

        if let Some(stderr) = child.stderr.take() {
            use std::io::{BufRead, BufReader};
            let reader = BufReader::new(stderr);
            for line in reader.lines().flatten() {
                if let Some(current) = parse_progress(&line) {
                    if ctx.total_secs > 0.0 {
                        let pct = (current / ctx.total_secs * 100.0).min(99.0);
                        let _ = ctx.app.emit("video-progress", pct);
                    }
                }
            }
        }

        let status = child.wait().map_err(|e| e.to_string())?;
        if !status.success() {
            return Err(format!("ffmpeg 执行失败, exit code: {:?}", status.code()));
        }

        let output_size = std::fs::metadata(ctx.output_path)
            .map(|m| m.len())
            .unwrap_or(0);
        let processing_time = start.elapsed().as_secs_f64();

        let _ = ctx.app.emit("video-progress", 100.0);

        Ok(AudioOperationResult {
            output_path: ctx.output_path.to_string(),
            original_size,
            output_size,
            saved_bytes: original_size.saturating_sub(output_size),
            processing_time_secs: processing_time,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_extract_copy_args_structure() {
        let args = build_extract_copy_args("/input.mp4", "/output.aac");
        assert_eq!(args[0], "-y");
        assert_eq!(args[1], "-i");
        assert_eq!(args[2], "/input.mp4");
        assert_eq!(args[3], "-vn");
        assert_eq!(args[4], "-c:a");
        assert_eq!(args[5], "copy");
        assert_eq!(args[6], "/output.aac");
    }

    #[test]
    fn test_build_extract_copy_args_no_bitrate() {
        let args = build_extract_copy_args("/in.mkv", "/out.m4a");
        // copy 模式不应该有 -b:a
        assert!(!args.contains(&"-b:a".to_string()));
        assert!(!args.contains(&"libmp3lame".to_string()));
    }
}
